### ISSUE TITLE
make lib tree-shakable

### ISSUE THREAD

**Issue by nonlinearcom** [MelihAltintas/vue3-openlayers#378]
Hi, after adding vue3-openlayers as a plugin to a clean Nuxt project, the client bundle shows a large file (vue3-openlayers.es.js) of over 1MB. The same issue occurs when performing selected imports, like `import { Map, Layers, Sources } from "vue3-openlayers` (975 KB). 

Is it possible to reduce the size of the bundled package? 🤔 
 
```
├── ol-contextmenu@5.5.0
├── ol-ext@4.0.23
├── ol@10.1.0
├── vue@3.4.38 invalid: "latest" from the root project
└── vue3-openlayers@11.1.0
```

**To Reproduce**

[Reproduction repo](https://github.com/nonlinearcom/vue3openlayers-nuxt)

Run `npx nuxi analyze`

**Screenshots**
<img width="1721" alt="Screenshot 2024-08-29 at 16 22 46" src="https://github.com/user-attachments/assets/5207e022-4bd8-4f3e-9864-049eb776c27a">

**Comment by YXYH-12138** [comment]
Yes, I have encountered the same issue, and this library cannot be tree-shaken.🤔

**Comment by d-koppenhagen** [comment]
Hey, would like to contribute and refactor it? I think it would be acceptable if a Breaking Change is introduced. The best in my opinion would be to allow to import / load components individually and therefore expose them.

**Comment by vincerubinetti** [comment]
Hmm this is a bit concerning.

I have a SPA Vue app using Vite. Using `vite-bundle-visualizer`, I get this:

<img width="1701" alt="Image" src="https://github.com/user-attachments/assets/6db9bfd5-86fe-4404-9bc7-2d25ee9d27ba" />

It looks like this package is taking up ~1.5 MB (uncompressed) and OL itself ~2 MB. In the OL breakdown, I _think_ those sub-assets are only ones I'm actually using (though I'm not using the webgl renderer as far as I know). Not sure how much of that 1.5 MB is necessary vs un-necessary.

Over-the-wire, both libraries together are about 750kb, which is pretty bad... but the library does provide a lot of utility...

**Comment by gastoneb** [comment]
Build size was a deal-breaker for us too, and I think it's important to note that other web mapping libraries have similar issues. For example, MapLibre-GL is about 890 kB. Thankfully ol is more modular, so in theory this should be avoidable! Nevertheless, with vue3-openlayers you currently end up with the full size of ol (695 kB for me), plus that large vue3-openlayers.es.js file (892 kB for me), plus ol-ext (141 kB for me) no matter how simple your app. 

I think the idea to expose individual components makes sense, even though it would be breaking (not to mention a fair bit of work when you factor in the documentation and examples). Do we think a single volunteer could take this all on themselves? I would also hope that this solution provides users a way to avoid bundling ol-ext. The ol-ext CSS file alone is 133 kB, nearly the size of Leaflet. And anyway, if you are using ol in a vue app there's a good chance you are writing your own custom controls. If that doesn't work, I would suggest potentially removing ol-ext as a dependency.

In terms of workarounds, it's fine to use plain ol in your application but of course you lose out on the benefits of Vue componentry. Another option is to write your own components using only the ol functionality that you need. We built some basic custom components similar to OlMap.vue, OlView.vue, OlTileLayer.vue, OlVectorLayer.vue, OlSourceXYZ.vue, and OlSourceVector.vue. In doing so, we got our build size down to 296 kB (from 695 + 892 + 141 = 1728 kB with vue3-openlayers); see below. The build size was only 2.9 kB larger than the same app built with plain ol, for the component code. Another potential strategy for anyone considering using vue3-openlayers might be to fork the repo and rip out everything you don't need, although I haven't tried it, and I don't know if it will still bundle all of ol and ol-ext. At least this way you preserve the same well-documented interface. 

Esbuild summary of deps for custom vue components for a minimal working map (xyz layer + vector):

![Image](https://github.com/user-attachments/assets/5f0aaf53-af57-47b1-ac2d-9f6b714c7871)

**Comment by d-koppenhagen** [comment]
I totally agree! I would be very happy If one can make a good suggestion for making this lib more modular and ideally contribute to this. I think I won't be able to do this by myself the next month. So if you want to enhance this lib: Let's do it! I think it will be totally acceptable if breaking changes will be introduced if it makes the whole lib more flexible and smaller. Let's me know how I can support you and I will do my best!
